### PR TITLE
DOCSP-30562: operation error handling

### PR DESCRIPTION
**Pull request by rustagir**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-rust/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-30562>
Staging:
- https://preview-mongodbrustagir.gatsbyjs.io/rust/DOCSP-30562-ops-error-handling/op-error-handling/
- https://preview-mongodbrustagir.gatsbyjs.io/rust/DOCSP-30562-ops-error-handling/#operation-error-handling

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST? > ignore the error, it is being handled in a different PR
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

had to manually run vale - fixed issues:
✔ 0 errors, 0 warnings and 0 suggestions in 1 file.

### CODE REVIEW

File: source/op-error-handling.txt [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+   :copyable: false
+
+   Error: Error { kind: Write(WriteError(WriteError { code: 11000,
+   code_name: None, message: "E11000 duplicate key error collection:
+   db.test_coll index: _id_ dup key: { _id: 1 }", details: None })), labels:
+   {}, wire_version: None, source: None }
+
+In the preceding error message, the value of the ``kind`` field is
+``Write``, which applies when an error occurs when
+you try to execute a write operation.
```
Comment (norareidy) [inline source/op-error-handling.txt:62]: S:

1. This sentence is a little wordy and repeats "when," you could change to something like:
"... the value of the ``kind`` field is ``Write``. This value applies when an error occurs during a write operation."

2. style guide is weird about "execute," could replace with "run" (or rephrased like the suggestion above)

File: source/op-error-handling.txt [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+   ..
+     - The :ref:`Frequently Asked Questions (FAQ) <golang-faq>` for the
+       {+driver-short+}
+     - :ref:`rust-connection-troubleshooting`
+
+Error Type
+----------
+
+Most of the exceptions that the driver throws when executing operations
```
Comment (norareidy) [inline source/op-error-handling.txt:39]: S: style guide prefers "generate" to "throw" (in the "Write for a Global Audience" section)

File: source/op-error-handling.txt [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+are of the `Error <{+api+}/error/struct.Error.html>`__ type. The
+``Error`` instance that an operation
+returns contains the ``kind`` field, which describes the type of error
+that occurred. The ``kind`` field has a value of type `ErrorKind
+<{+api+}/error/enum.ErrorKind.html>`__, which has many variants for
+different kinds of errors.
```
Comment (norareidy) [inline source/op-error-handling.txt:45]: S: it might be worthwhile to include a list of some of these variants, just because "many variants for different kinds of errors" seems pretty vague

File: source/op-error-handling.txt [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+In the preceding error message, the value of the ``kind`` field is
+``Write``, which applies when an error occurs when
+you try to execute a write operation.
+
+Connection Errors
+-----------------
+
+You might successfully connect to a MongoDB server, but the connection
+pool might be cleared due to a concurrent operation error. In this
```
Comment (norareidy) [inline source/op-error-handling.txt:68]: S: instead of using "might" as a helping verb twice, you can reword this to make the verbs a little simpler:
"A concurrent operation error might interrupt a successful connection to a MongoDB server and clear the connection pool"

File: source/op-error-handling.txt [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+takes a value of type `WriteError <{+api+}/error/struct.WriteError.html>`__ or
+`WriteConcernError <{+api+}/error/struct.WriteConcernError.html>`__.
+
+Write Concern Error
+~~~~~~~~~~~~~~~~~~~
+
+The driver raises a ``WriteConcernError`` error when you execute a write
+operation and the driver cannot satisfy the specified write concern. For
+example, if you specify a write concern of ``majority`` for
+operations on a replica set with three nodes, the driver would return
```
Comment (norareidy) [inline source/op-error-handling.txt:103]:
```suggestion
operations on a replica set with three nodes, the driver returns
```

File: source/op-error-handling.txt [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+it raises an error where the ``kind`` field value is ``Write``.
+The body of the ``Write`` variant is the enum `WriteFailure
+<{+api+}/error/enum.WriteFailure.html>`__, which
+takes a value of type `WriteError <{+api+}/error/struct.WriteError.html>`__ or
+`WriteConcernError <{+api+}/error/struct.WriteConcernError.html>`__.
+
+Write Concern Error
+~~~~~~~~~~~~~~~~~~~
+
+The driver raises a ``WriteConcernError`` error when you execute a write
```
Comment (norareidy) [inline source/op-error-handling.txt:100]:
```suggestion
The driver raises a ``WriteConcernError`` error when you run a write
```

File: source/op-error-handling.txt [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+In the preceding error message, the ``message`` field describes the
+reason for the error, and the ``details`` field provides specific
+details about the failing operation.
+
+Additional Information
+----------------------
+
+To learn more about write concerns, see :manual:`Write Concern
+</reference/write-concern/>` in the Server manual.
+
```
Comment (norareidy) [inline source/op-error-handling.txt:142]: S: you link to some API documentation on this page (Error, ErrorKind, etc) that could be good to add here

Comment (norareidy) [reply source/op-error-handling.txt:142]: actually I'm not sure if we usually include API links in a separate section that are also linked in the text, so disregard if not

File: source/op-error-handling.txt [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+Overview
+--------
+
+This page describes errors you might encounter when
+using the {+driver-long+} to perform MongoDB operations.
+
+.. note::
+
+   This page addresses only operation error handling. If you encounter
```
Comment (norareidy) [inline source/op-error-handling.txt:21]:
```suggestion
   This page only addresses operation error handling. If you encounter
```

**Review by norareidy — CHANGES_REQUESTED** [review]
Great work, just a few suggestions from me!

File: source/op-error-handling.txt (hunk 1 above)
Comment (rustagir) [reply source/op-error-handling.txt:62]: reworded

File: source/op-error-handling.txt (hunk 6 above)
Comment (rustagir) [reply source/op-error-handling.txt:100]: changed all to perform

File: source/op-error-handling.txt (hunk 7 above)
Comment (rustagir) [reply source/op-error-handling.txt:142]: For this kind of page (not a fundamentals page), I think linking inline could be helpful. What do you also think about removing the additional information section and moving the content that's currently there to the places where they are more relevant?

File: source/op-error-handling.txt (hunk 8 above)
Comment (rustagir) [reply source/op-error-handling.txt:21]: Per the[ style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#o:~:text=Don%27t%20hyphenate.-,only,-Place%20only%20preceding), the only is currently in the correct place

Comment (norareidy) [reply source/op-error-handling.txt:23]: Oh good to know, sorry about that!

File: source/op-error-handling.txt [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+:manual:`Tuning Your Connection Pool Settings
+</tutorial/connection-pool-performance-tuning/>` in the Server manual.
+
+Depending on the circumstances that produce the error, the driver may
+add a ``RetryableWriteError`` label to the error, as shown in the
+following error message:
+
+.. code-block:: none
+   :copyable: false
+   :emphasize-lines: 1
```
Comment (norareidy) [inline source/op-error-handling.txt:91]: S: I'm not sure if this first line needs to be highlighted -- since you describe the RetryableWriteError label before and after the code block, maybe you could highlight line 3 instead (or remove the highlighting). Not a necessary change though if you want to emphasize the kind field

File: source/op-error-handling.txt (hunk 7 above)
Comment (norareidy) [reply source/op-error-handling.txt]: That sounds good to me!

**Review by norareidy — APPROVED** [review]
LGTM with one suggestion!

File: source/op-error-handling.txt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+   ..
+     - The :ref:`Frequently Asked Questions (FAQ) <golang-faq>` for the
+       {+driver-short+}
+     - :ref:`rust-connection-troubleshooting`
+
+Error Type
+----------
+
+Most of the exceptions that the driver generates when performing operations
+are of the `Error <{+api+}/error/struct.Error.html>`__ type. The
```
Comment (kevinAlbs) [inline source/op-error-handling.txt:42]:
```suggestion
Errors returned by the driver when performing operations are of the `Error <{+api+}/error/struct.Error.html>`__ type. The
```

[`Result` documents](https://github.com/mongodb/mongo-rust-driver/blob/6a9ef26bd1adc721fb59e61a6895d95cef3e3200/src/error.rs#L37):
> The result type for all methods that can return an error in the `mongodb` crate.

I expect all errors are returned as `Error`.

Suggest avoiding the word "exception", since Rust does not have exceptions.

**Review by kevinAlbs — APPROVED** [review]
LGTM with suggested change.
